### PR TITLE
MULTIARCH-4762: Add e2e for pod nodeName exist

### PR DESCRIPTION
**Pull request by lwan-wanglin**
Add e2e for pod nodeName exist

### CODE REVIEW

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 1]
```diff
@@ -921,6 +945,27 @@ func getImageConfig(ctx context.Context, client runtimeclient.Client) ocpconfigv
 	return image
 }
 
+func getNodeName(labelKey string, labelInValue string) string {
+	var (
+		r   *labels.Requirement
+		err error
+	)
+	if labelInValue == "" {
+		r, err = labels.NewRequirement(labelKey, "Exists", nil)
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go]: for `"Exists`" and `"in"`, we should have constant in the api machinery or some other library.

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 2]
```diff
@@ -921,6 +945,27 @@ func getImageConfig(ctx context.Context, client runtimeclient.Client) ocpconfigv
 	return image
 }
 
+func getNodeName(labelKey string, labelInValue string) string {
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go]: can we add this to pkg/testing/framework/node.go and call it `getRandomNodeName`?

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 3]
```diff
@@ -254,6 +254,30 @@ var _ = Describe("The Pod Placement Operand", func() {
...
+				WithSelectorAndPodLabels(podLabel).
+				WithPodSpec(ps).
+				WithReplicas(utils.NewPtr(int32(1))).
+				WithName("test-deployment").
+				WithNamespace(ns.Name).
+				Build()
+			err = client.Create(ctx, &d)
+			Expect(err).NotTo(HaveOccurred())
+			verifyPodNodeAffinity(ns, "app", "test")
+			verifyPodLabels(ns, "app", "test", e2e.Absent, schedulingGateLabel)
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go:284]: I would verify more that the pod is running

File: pkg/e2e/podplacement/pod_placement_test.go (hunk 1 above)
Comment (lwan-wanglin) [reply pkg/e2e/podplacement/pod_placement_test.go]: ok

File: pkg/e2e/podplacement/pod_placement_test.go (hunk 2 above)
Comment (lwan-wanglin) [reply pkg/e2e/podplacement/pod_placement_test.go]: sute, will fix it.

File: pkg/e2e/podplacement/pod_placement_test.go (hunk 3 above)
Comment (lwan-wanglin) [reply pkg/e2e/podplacement/pod_placement_test.go:284]: thanks, that's a good point.

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 4]
```diff
@@ -254,6 +255,33 @@ var _ = Describe("The Pod Placement Operand", func() {
 			verifyPodNodeAffinity(ns, "app", "test")
 			verifyPodLabels(ns, "app", "test", e2e.Present, schedulingGateLabel)
 		})
+		It("should not set the node affinity when nodeName exist", func() {
+			var err error
+			ns := framework.NewEphemeralNamespace()
+			err = client.Create(ctx, ns)
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go:261]:
```suggestion
			By("Create an ephemeral namespace")
			err = client.Create(ctx, ns)
```

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 5]
```diff
@@ -254,6 +255,33 @@ var _ = Describe("The Pod Placement Operand", func() {
...
+			Expect(err).NotTo(HaveOccurred())
+			//nolint:errcheck
+			defer client.Delete(ctx, ns)
+			workerNodeName, err := framework.GetRandomNodeName(ctx, client, "node-role.kubernetes.io/worker", "")
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go:266]:
```suggestion
  
			By("Get the name of a random worker node")
			workerNodeName, err := framework.GetRandomNodeName(ctx, client, "node-role.kubernetes.io/worker", "")
```

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 6]
```diff
@@ -1023,6 +1055,24 @@ func verifyPodLabels(ns *corev1.Namespace, labelKey string, labelInValue string,
 	}, e2e.WaitShort).Should(Succeed())
 }
 
+func verifyPodsAreRunning(ns *corev1.Namespace, labelKey string, labelInValue string) {
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go]: Can this stay in a proper file at pkg/testing. I would start having this 'verify' methods in a proper file and decoupled from the BDD logic of the test cases.

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 7]
```diff
@@ -254,6 +255,37 @@ var _ = Describe("The Pod Placement Operand", func() {
...
+			ns := framework.NewEphemeralNamespace()
+			err = client.Create(ctx, ns)
+			Expect(err).NotTo(HaveOccurred())
+			//nolint:errcheck
+			defer client.Delete(ctx, ns)
+			By("Get the name of a random worker node")
+			workerNodeName, err := framework.GetRandomNodeName(ctx, client, "node-role.kubernetes.io/worker", "")
+			Expect(workerNodeName).NotTo(BeEmpty())
+			Expect(err).NotTo(HaveOccurred())
+			By("Create a deployment using the container with nodeName existing")
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go]:
```suggestion
			By("Create a deployment using the container with a preset nodeName")
```

File: pkg/e2e/podplacement/pod_placement_test.go [hunk 8]
```diff
@@ -254,6 +255,37 @@ var _ = Describe("The Pod Placement Operand", func() {
...
+				WithReplicas(utils.NewPtr(int32(1))).
+				WithName("test-deployment").
+				WithNamespace(ns.Name).
+				Build()
+			err = client.Create(ctx, &d)
+			Expect(err).NotTo(HaveOccurred())
+			By("Check PPO should not add schedulingGate label on pod and pod can be created and running")
+			verifyPodsAreRunning(ns, "app", "test")
+			verifyPodNodeAffinity(ns, "app", "test")
+			verifyPodLabels(ns, "app", "test", e2e.Absent, schedulingGateLabel)
```
Comment (aleskandro) [inline pkg/e2e/podplacement/pod_placement_test.go]:
```suggestion
			By("The pod should not have been processed by the webhook and the scheduling gate label should not be added")
			verifyPodLabels(ns, "app", "test", e2e.Absent, schedulingGateLabel)
			By("The pod should keep the same node affinity provided by the users. No node affinity is added by the controller.")
			verifyPodNodeAffinity(ns, "app", "test")
			By("The pod should be running and not gated")
			verifyPodsAreRunning(ns, "app", "test")
```

I don't like them 100%, but having more detailed comments in the verification steps in particular is nice as it's easier to find which step is failing precisely.

**Review by aleskandro — COMMENTED** [review]
Thanks @lwan-wanglin just a few more comments, but overall, it's good.

File: pkg/e2e/podplacement/pod_placement_test.go (hunk 6 above)
Comment (lwan-wanglin) [reply pkg/e2e/podplacement/pod_placement_test.go]: ok, will fix it.

**Comment by lwan-wanglin** [comment]
@aleskandro could you help to review again? thanks!

**Comment by aleskandro** [comment]
Thanks @lwan-wanglin

**Pull request merged**